eth/gaspricemonitor.go

```diff
@@ -139,3 +148,109 @@ func (gpm *GasPriceMonitor) updateGasPrice(gasPrice *big.Int) {
 
 	gpm.gasPrice = gasPrice
 }
+
+type gasPrices struct {
+	sync.RWMutex
+	list   []*big.Int
+	length int
+}
+
+func newGasPrices(length int) *gasPrices {
+	return &gasPrices{
+		length: length,
+		list:   make([]*big.Int, 0),
+	}
+}
+
+func (q *gasPrices) front() *big.Int {
+	q.RLock()
+	defer q.RUnlock()
+	if len(q.list) == 0 {
+		return nil
+	}
+	return q.list[0]
+}
+
+func (q *gasPrices) back() *big.Int {
+	q.RLock()
+	defer q.RUnlock()
+	if len(q.list) == 0 {
+		return nil
+	}
+	return q.list[len(q.list)-1]
+}
+
+func (q *gasPrices) isFull() bool {
+	q.RLock()
+	defer q.RUnlock()
+	return len(q.list) == q.length
+}
+
+func (q *gasPrices) push(price *big.Int) {
+	isFull := q.isFull()
+
+	q.Lock()
+	defer q.Unlock()
+
+	// Queue is not full, simply append to back
+	if !isFull {
+		q.list = append(q.list, price)
+		return
+	}
+
+	list := q.list[1:]
+	list = append(list, price)
+	q.list = list
+}
+
+func (q *gasPrices) mean() *big.Int {
+	q.RLock()
+	defer q.RUnlock()
+	sum := big.NewInt(0)
+
+	if len(q.list) == 0 {
+		return sum
+	}
+
+	for _, v := range q.list {
+		sum = sum.Add(sum, v)
+	}
+	return sum.Div(sum, big.NewInt(int64(len(q.list))))
+}
+
+func (q *gasPrices) isOutlier(price *big.Int) bool {
+	q.RLock()
+	defer q.RUnlock()
+
+	if len(q.list) <= 1 {
+		return false
+	}
+
+	var (
+		variance   = big.NewInt(0)
+		length     = big.NewInt(int64(len(q.list)))
+		deviations = big.NewInt(3) // The max acceptable std, anything outside will be removed from the set
+	)
+
+	mean := q.mean()
+	// Calculate variance (sum(x - mean)^2 )/ length
+	for _, price := range q.list {
+		// Calculate the summation
+		x := new(big.Int).Sub(price, mean)
```

Comment:
If all the prices in the list are the same, then x = 0 and the variance = 0. As a result, the upper/lower bound would both be the mean so all prices would be discarded unless they are exactly equal to the mean.

Reply:
We don't insert a price into the list if it is the same as the back of the queue to avoid this exact scenario https://github.com/livepeer/go-livepeer/blob/3a9637e0c5513bc64a94d6779acb3116a594490c/eth/gaspricemonitor.go#L131